velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,269 @@
+  auto rawOffsets = baseArray->rawOffsets();
+  auto rawNulls = baseArray->rawNulls();
+  auto rawChildHashes = childBuffer->as<int64_t>();
+  auto hashBuffer = buffer->asMutable<int64_t>();
+
+  rows.applyToSelected([&](auto row) {
+    int64_t hash = 0;
+    if (!(rawNulls && bits::isBitNull(rawNulls, row))) {
+      auto size = rawSizes[indices[row]];
+      auto offset = rawOffsets[indices[row]];
+
+      for (int i = 0; i < size; i++) {
+        hash = 31 * hash + rawChildHashes[offset + i];
+      }
+    }
+    hashBuffer[row] = hash;
+  });
+
+  // Free the buffer ?
+  childBuffer->release();
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::MAP>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  auto baseMap = vector_->base()->as<MapVector>();
+  VELOX_CHECK_EQ(children_.size(), 2)
+
+  auto convertedRows = velox::functions::toElementRows(
+      baseMap->mapKeys()->size(), rows, baseMap, vector_->indices());
+  BufferPtr keyBuffer =
+      AlignedBuffer::allocate<int64_t>(convertedRows.end(), baseMap->pool());
+
+  BufferPtr valueBuffer =
+      AlignedBuffer::allocate<int64_t>(convertedRows.end(), baseMap->pool());
+
+  children_[0]->bulkHash(convertedRows, keyBuffer);
+  children_[1]->bulkHash(convertedRows, valueBuffer);
+
+  auto keyHashes = keyBuffer->as<int64_t>();
+  auto valueHashes = valueBuffer->as<int64_t>();
+  auto hashBuffer = buffer->asMutable<int64_t>();
+
+  auto indices = vector_->indices();
+  auto rawSizes = baseMap->rawSizes();
+  auto rawOffsets = baseMap->rawOffsets();
+  auto rawNulls = baseMap->rawNulls();
+
+  rows.applyToSelected([&](auto row) {
+    int64_t hash = 0;
+    if (!(rawNulls && bits::isBitNull(rawNulls, row))) {
+      auto size = rawSizes[indices[row]];
+      auto offset = rawOffsets[indices[row]];
+
+      for (int i = 0; i < size; i++) {
+        hash = keyHashes[offset + i] ^ valueHashes[offset + i];
+      }
+    }
+    hashBuffer[row] = hash;
+  });
+
+  keyBuffer->release();
+  valueBuffer->release();
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::ROW>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  auto baseRow = vector_->base()->as<RowVector>();
+
+  BufferPtr rowHashBuffer =
+      AlignedBuffer::allocate<int64_t>(rows.end(), baseRow->pool());
+
+  auto hashBuffer = buffer->asMutable<int64_t>();
+  auto rowHashes = rowHashBuffer->as<int64_t>();
+
+  for (int i = 0; i < rows.end(); i++) {
+    hashBuffer[i] = 1;
+  }
+
+  for (int i = 0; i < baseRow->childrenSize(); i++) {
+    children_[i]->bulkHash(rows, rowHashBuffer);
+
+    rows.applyToSelected([&](auto row) {
+      hashBuffer[row] = 31 * hashBuffer[row] + rowHashes[row];
+    });
+  }
+
+  rowHashBuffer->release();
+}
+
+// BufferPtr is modified and results of hash are stored there.
+void PrestoHasher::bulkHash(const SelectivityVector& rows, BufferPtr& buffer) {
+  auto kind = vector_->base()->typeKind();
+  VELOX_DYNAMIC_TYPE_DISPATCH(bulkHash, kind, rows, buffer);
```

Comment:
bulkHash template is defined separately for each template type, hence, there seems not much benefit from using VELOX_DYNAMIC_TYPE_DISPATCH. A simple switch statement might be clearer.